ouroboros-network/test/Test/Ouroboros/Network/Testnet.hs

```diff
@@ -851,9 +860,291 @@ prop_diffusion_ig_valid_transition_order defaultBearerInfo diffScript =
+
+        events :: [Trace () (Time, DiffusionTestTrace)]
+        events = fmap ( Trace.fromList ()
+                      . fmap (\(WithName _ (WithTime t b)) -> (t, b)))
+               . Trace.toList
+               . splitWithNameTrace
+               . Trace.fromList ()
+               . fmap snd
+               . Signal.eventsToList
+               . Signal.eventsFromListUpToTime (Time (10 * 60 * 60))
+               . Trace.toList
+               . fmap (\(WithTime t (WithName name b))
+                       -> (t, WithName name (WithTime t b)))
+               . withTimeNameTraceEvents
+                  @DiffusionTestTrace
+                  @NtNAddr
+               $ runSimTrace sim
+
+     in conjoin
+      $ verify_timeouts
+      <$> events
+
+  where
+    verify_timeouts :: Trace () (Time, DiffusionTestTrace) -> Property
+    verify_timeouts events =
+      let transitionSignal :: Trace () [(Time, AbstractTransitionTrace NtNAddr)]
+          transitionSignal = groupConns snd
+                           . selectDiffusionConnectionManagerTransitionEventsTime
+                           $ events
+
+       in getAllProperty
+        $ verifyAllTimeouts transitionSignal
+       where
+         verifyAllTimeouts :: Trace () [(Time, AbstractTransitionTrace NtNAddr)]
+                           -> AllProperty
+         verifyAllTimeouts =
+           bifoldMap
+            (const mempty)
+            (\ tr ->
+              AllProperty
+              $ counterexample ("\nConnection transition trace:\n"
+                              ++ intercalate "\n" (map show tr)
+                              )
+              $ verifyTimeouts Nothing tr)
+
+         -- verifyTimeouts checks that in all \tau transition states the timeout is
+         -- respected. It does so by checking the stream of abstract transitions
+         -- paired with the time they happened, for a given connection; and checking
+         -- that transitions from \tau states to any other happens withing the correct
+         -- timeout bounds. One note is that for the example
+         -- InboundIdleState^\tau -> OutboundState^\tau -> OutboundState sequence
+         -- The first transition would be fine, but for the second we need the time
+         -- when we transitioned into InboundIdleState and not OutboundState.
+         --
+         verifyTimeouts :: Maybe (AbstractState, Time)
+                        -- ^ Map of first occurence of a given \tau state
+                        -> [(Time , AbstractTransitionTrace NtNAddr)]
+                        -- ^ Stream of abstract transitions for a given connection
+                        -- paired with the time it ocurred
+                        -> Property
+         -- In Diffusion tests there can be truncated transitions
+         verifyTimeouts _ [] = property True
+         -- If we already seen a \tau transition state
+         verifyTimeouts st@(Just (state, t')) ((t, TransitionTrace _ tt@(Transition _ to)):xs) =
```

Comment:
We can not share it because it is defined in the tests package of `ouroboros-network-framework` if that was exported then we could re-use it fine. I am not aware of the implications of doing that though.

Reply:
Let's move it to `ouroboros-network-framework:testlib`.

Reply:
Done!

Reply:
No, I forgot I had it in 2 places and forgot to remove this one!